src/time/ext.rs

```diff
@@ -0,0 +1,298 @@
+//! Extensions for Futures types.
+
+use std::future::Future;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+
+use futures::{AsyncRead, Stream};
+
+use super::Delay;
+
+/// A future returned by methods in the [`FutureTimeExt`] trait.
+///
+/// [`FutureTimeExt.timeout`]: trait.FutureTimeExt.html
+#[derive(Debug)]
+pub struct Timeout<F: Future> {
+    future: F,
+    delay: Delay,
+}
+
+impl<F: Future> Timeout<F> {
+    pin_utils::unsafe_pinned!(future: F);
+    pin_utils::unsafe_pinned!(delay: Delay);
```

Comment:
I'm quite unhappy with using `unsafe_pinned` here. I copied the approach from the `futures-timer` crate, but I suspect we could probably do without it. Suggestions would be much appreciated!